### PR TITLE
[MFAC] training aware implementation for GradSampler

### PR DESCRIPTION
**Pull request by bfineran**
* updates the `GradSampler` class to allow for more flexible inputs to models with `args` and `kwargs`
* allows for `GradSampler`s to be passed on recipe initialization so that they may be run before pruning steps to avoid the need for gradient buffering during training. if no grad samplers are passed, normal gradient buffering will occur
* will need to do some small live tests, and then get an integration with our BERT flows

### CODE REVIEW

File: research/mfac/tutorials/one_shot_pruning_with_mfac.md [hunk 1]
```diff
@@ -118,11 +131,12 @@ data_loader = DataLoader(dataset, shuffle=True, batch_size=16)
...
     for sample in data_loader:
-        yield [t.to(device) if isinstance(t, torch.Tensor) else t for t in sample]
+        img, target = [t.to(device) if isinstance(t, torch.Tensor) else t for t in sample]
+        yield [img], {}, target
 
 # wrap cross entropy loss function to use correct output
-def loss_function(x, y):
-    return torch.nn.functional.cross_entropy(x[0], y)
+def loss_function(model_outputs, loss_target):
+    return torch.nn.functional.cross_entropy(model_outputs[0], loss_target)
```
Comment (spacemanidol) [inline research/mfac/tutorials/one_shot_pruning_with_mfac.md:139]: shouldn't we have an assert here that model_outputs is a tuple? Also worth adding a check that the shapes of loss target and model outputs match right?

**Review by spacemanidol — APPROVED** [review]
Looks good. Minor comment about adding an assert on outputs vs target outputs but feel free to ignore

**Review by rahul-tuli — APPROVED** [review]
